src/models/circle_packer/CirclePacker.js

```diff
@@ -0,0 +1,357 @@
+        }
+      }
+    }
+
+    return valid ? possibleC : null
+  }
+
+  growCircles() {
+    if (this.circles.length > 1) {
+      while (this.circles.filter(circle => circle.growing).length > 0) {
+        this.circles.forEach(c => {
+          if (c.growing) {
+            if (c.outOfBounds()) {
+              c.growing = false
+            } else {
+              this.circles.forEach(other => {
+                if (c !== other) {
+                  let d = c.distance(other)
+                  if (d < c.r + other.r) {
+                    c.growing = false
+                    this.graph.addEdge(c, other)
+                  }
+                }
+              })
+            }
+
+            c.grow()
+          }
+        })
+      }
+    }
+  }
+
+  markConnected(c) {
+    c.connected = true
+    const neighbors = this.graph.neighbors(c)
+
+    for (const curr of neighbors) {
+      if (!curr.connected) {
+        this.markConnected(curr)
+      }
+    }
+  }
+
+  drawCircle(c, angle) {
+    if (!c.drawn) {
+      this.points.push(...circle(c.r, angle, c.x, c.y))
+      c.drawn = true
+    }
+  }
+
+  drawConnectingArc(from, to, startingAngle) {
+    const li1 = this.closest(from.lineIntersection(from, to), to)
+    const li2 = this.closest(to.lineIntersection(from, to), from)
+    const a2 = Math.atan2(li1.y - from.y, li1.x - from.x)
+    const a3 = Math.atan2(li2.y - to.y, li2.x - to.x)
+
+    this.points.push(...arc(from.r, startingAngle, a2, from.x, from.y))
+    return a3
+  }
+
+  walk(c, angle, stack) {
+    const neighbors = this.graph.neighbors(c)
+    let a1 = angle
+
+    c.walked = true
+    stack.unshift(c)
+
+    for (const curr of neighbors) {
+      if (!curr.walked) {
+        const a2 = this.drawConnectingArc(c, curr, a1)
+        a1 = this.walk(curr, a2, stack)
+      }
+    }
+
+    this.points.push(...circle(c.r, a1, c.x, c.y))
+    return this.walkBack(stack) || a1
+  }
+
+  walkBack(stack) {
+    const c = stack.shift()
+    let a1, a2
+
+    if (stack.length > 0) {
+      const p1 = stack[0]
+      const li1 = this.closest(c.lineIntersection(c, p1), c)
+      a1 = Math.atan2(li1.y - p1.y, li1.x - p1.x)
+
+      if (stack.length > 1) {
+        const p2 = stack[1]
+        const li2 = this.closest(p1.lineIntersection(p1, p2), p2)
+
+        a2 = Math.atan2(li2.y - p1.y, li2.x - p1.x)
+        this.points.push(...arc(p1.r, a1, a2, p1.x, p1.y))
+
+        return a2
+      } else {
+        return a1
+      }
+    }
+  }
+
+  connectAlongPerimeter(c, end, angle) {
+    debugger
+    const intersection = this.boundaryIntersection(c)
+    const intersection2 = this.boundaryIntersection(end)
+
+    // draw arc to perimeter
+    const a2 = Math.atan2(intersection.y - c.y, intersection.x - c.x)
+    this.points.push(...arc(c.r, angle, a2, c.x, c.y))
+
+    // connect along perimeter
+    const machine = getMachineInstance([], this.settings)
+    this.points.push(...machine.tracePerimeter(intersection, intersection2))
```

Comment:
I found a case where it is tracing across the pattern. It looks like this is where that would be coming from, but I'm not sure what is causing the issue.

![screenshot-2021-09-01-1630513422](https://user-images.githubusercontent.com/1812882/131709613-1deceef2-9934-4ec4-a977-c9eb15023e1a.png)

This is the default 0-500x0-500 rectangle machine.

Reply:
Maybe we could make these perimeter points a lot larger? I'm not sure why it is trying to go around to that other corner before tracking on the circle to the left.

Reply:
It's not this line of code. This line is fine. I added some logs and these lines aren't causing the issue.

Reply:
This is fixed by #214. I'm going to merge this PR now.